writer_test.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/uber-common/zap/spy"
+)
+
+type testBuffer struct {
+	*bytes.Buffer
+}
+
+func newTestBuffer() testBuffer {
+	return testBuffer{&bytes.Buffer{}}
+}
+
+func (t testBuffer) Sync() error {
+	return nil
+}
+
+func requireWriteWorks(t testing.TB, ws WriteSyncer) {
+	n, err := ws.Write([]byte("foo"))
+	require.NoError(t, err, "Unexpected error writing to WriteSyncer.")
+	require.Equal(t, 3, n, "Wrote an unexpected number of bytes.")
+}
+
+func TestAddSyncWriteSyncer(t *testing.T) {
+	buf := &bytes.Buffer{}
+	concrete := &spy.WriteSyncer{Writer: buf}
+	ws := AddSync(concrete)
+	requireWriteWorks(t, ws)
+
+	require.NoError(t, ws.Sync(), "Unexpected error syncing a WriteSyncer.")
+	require.True(t, concrete.SyncCalled, "Expected to dispatch to concrete type's Sync method.")
+
+	concrete.Err = errors.New("fail")
+	assert.Error(t, ws.Sync(), "Expected to propagate errors from concrete type's Sync method.")
+}
+
+func TestAddSyncWriteFlusher(t *testing.T) {
+	buf := &bytes.Buffer{}
+	concrete := &spy.WriteFlusher{Writer: buf}
+	ws := AddSync(concrete)
+	requireWriteWorks(t, ws)
+
+	require.NoError(t, ws.Sync(), "Unexpected error syncing a WriteSyncer.")
+	require.True(t, concrete.FlushCalled, "Expected to dispatch to concrete type's Flush method.")
+
+	concrete.Err = errors.New("fail")
+	assert.Error(t, ws.Sync(), "Expected to propagate errors from concrete type's Flush method.")
+}
+
```

Comment:
let's put a test for `AddSync` where there's a `Sync` and a `Flush` method -- only the `Sync` should be called, but just want to have it tested.